File: shared/types/game.ts
export type Platform = 'bilibili' | 'douyin' | 'kuaishou'

export type Difficulty = 'easy' | 'medium' | 'hard'

export interface WordPuzzle {
  id: string
  word: string
  wordLength: number
  category: string
  difficulty: Difficulty
}

export interface GuessRecord {
  userId: string
  userName: string
  guess: string
  affinity: number
  timestamp: number
}

export interface Player {
  userId: string
  userName: string
  totalScore: number
  currentScore: number
  guessCount: number
}

export interface RoomState {
  roomId: string
  platform: Platform
  currentPuzzle: WordPuzzle | null
  streak: number
  highestAffinity: number
  guessBoard: GuessRecord[]
  leaderboard: Player[]
  previousPuzzle: string | null
  solvedBy: string | null
  revealedChars: boolean[]
}

export interface RoomConfig {
  roomId: string
  platform: Platform
  autoNextDelay: number
}

export interface PuzzleSolvedEvent {
  word: string
  solvedBy: string
}

export interface SocketEvents {
  'room:join': { roomId: string; platform: Platform; clientId: string }
  'room:leave': { roomId: string }
  'game:newPuzzle': WordPuzzle
  'game:guess': { roomId: string; userId: string; userName: string; guess: string; clientId: string }
  'game:guessResult': GuessRecord
  'game:state': RoomState
  'game:leaderboard': Player[]
  'game:previousPuzzle': string
  'game:puzzleSolved': PuzzleSolvedEvent
}

File: backend/app/models/game.py
from __future__ import annotations
from pydantic import BaseModel, ConfigDict
from pydantic.alias_generators import to_camel
from typing import Literal, Optional

Platform = Literal["bilibili", "douyin", "kuaishou"]
Difficulty = Literal["easy", "medium", "hard"]


class CamelModel(BaseModel):
    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
    )


class WordPuzzle(CamelModel):
    id: str
    word: str
    word_length: int
    category: str
    difficulty: Difficulty


class GuessRecord(CamelModel):
    user_id: str
    user_name: str
    guess: str
    affinity: float
    timestamp: int


class Player(CamelModel):
    user_id: str
    user_name: str
    total_score: int = 0
    current_score: int = 0
    guess_count: int = 0


class RoomState(CamelModel):
    room_id: str
    platform: Platform
    current_puzzle: Optional[WordPuzzle] = None
    streak: int = 0
    highest_affinity: float = 0.0
    guess_board: list[GuessRecord] = []
    leaderboard: list[Player] = []
    previous_puzzle: Optional[str] = None
    solved_by: Optional[str] = None
